### PR TITLE
Shared reader refactor

### PR DESCRIPTION
**Pull request by John-Leitch**
Sending the `SharedReader` refactor as well as some pretty extensive unit test coverage that uses XUnit's `Theory` to cover all primitives with a large number of corner cases. Perhaps a little over the top, but it sucks to have subtle issues in foundational types like this. Speaking of, that's actually why this PR is marked as draft. It turns out I had a bug in my refactor of `SharedReader.ReadChar`. I had used `Marshal.SizeOf<T>` in place of the `sizeof(T)` expressions, which returns a different value when passed `char` (`Marshal` returns `1` while `sizeof` returns `2`). Not to mention `Marshal` is intended for interop, so not exactly appropriate.

I've fixed this, but it made me realize there's a subtle issue in `ReadChar`: the current implementation always increases the offset by `2`, when the actual size of the `char` is based on the `Encoder` used by the underlying `BinaryReader`. By default utf8 is used, which is variable width. We could use `BaseStream.Position` to keep track of the actual size of the `char` read, but given the purpose of this type, it makes me wonder if `ReadChar` (or any of the `BinaryReader` encoder influenced methods) is even suitable. I see you already have some extensions for reading strings; allowing callers to dictate how the bytes are interpreted seems like a better approach here.

Finally, I've added overrides for most of the remaining `BinaryReader` functions to throw `NotImplementedException` in case of accidental use, as these would not update `Offset` and could cause subtle errors. The only one of note not overridden is `BinaryReader.Read()`, as `BinaryReader.ReadChar()` calls into this. Not really sure how to handle that one until a call on `ReadChar` is made.

### CODE REVIEW

**Comment by John-Leitch** [comment]
Refactored `SharedReader`, added in the `BaseStream.Position` size calculation as discussed. Added some tests, and during realized another bug: `ReadBytes` and similar may actually not return the number of bytes requested. Using `Position` remediates this, but added test coverage to be safe. We should be at 100% coverage for `SharedReader` now. Marking as ready for review.

File: Reemit.Common.UnitTests/SharedReaderTests.cs [hunk 1]
```diff
@@ -14,10 +18,200 @@ public void ReadBytes_Called_AdvancesRelativeOffsetAndOffset()
...
         
         // Act
-        sharedReader.ReadBytes(readBytesCount);
+        var actualBytes = sharedReader.ReadBytes(readBytesCount);
         
         // Assert
         Assert.Equal(sharedReaderOffset + readBytesCount, sharedReader.Offset);
         Assert.Equal(readBytesCount, sharedReader.RelativeOffset);
+        Assert.Equal(readBytesCount, actualBytes.Length);
+        Assert.NotNull(actualBytes);
```
Comment (v0idzz) [inline Reemit.Common.UnitTests/SharedReaderTests.cs]: It seems like the `NotNull` assertion on `actualBytes` may be redundant. We have an equality assertion just below it comparing `actualBytes` to the expected result. This already asserts that the value is not null.

File: Reemit.Common.UnitTests/SharedReaderTests.cs [hunk 2]
```diff
@@ -14,10 +18,200 @@ public void ReadBytes_Called_AdvancesRelativeOffsetAndOffset()
...
+
+        var expectedBytes = expectedUnmanagedValues.SelectMany(getBytes).ToArray();
+        var bytes = new byte[sharedReaderOffset].Concat(expectedBytes).ToArray();
+
+        using var stream = new MemoryStream(bytes);
+        using var binaryReader = new BinaryReader(stream);
+        using var sharedReader = new SharedReader(sharedReaderOffset, binaryReader, new object());
+
+        // Act
+        var actualUnmanagedValues = new T[expectedUnmanagedValues.Length];
```
Comment (v0idzz) [inline Reemit.Common.UnitTests/SharedReaderTests.cs:84]: Nitpicking here, but I believe that this step belongs to the `Arrange` section of the unit test. My personal take on this is that the `Act` section should only contain invocation of the method being tested. I see why we include the `for` loop in `Act`, but I think the array initialisation code can be moved to `Arrange`.

**Review by v0idzz — APPROVED** [review]
Good job, thanks. Left two comments, but it's up to you whether you want to address them as they aren't anything more than me nitpicking. Feel free to merge this PR as it is.

File: Reemit.Common.UnitTests/SharedReaderTests.cs (hunk 1 above)
Comment (John-Leitch) [reply Reemit.Common.UnitTests/SharedReaderTests.cs]: Done.

File: Reemit.Common.UnitTests/SharedReaderTests.cs (hunk 2 above)
Comment (John-Leitch) [reply Reemit.Common.UnitTests/SharedReaderTests.cs:84]: And done.
